File: Project_1_Java/src/PLD/classes/Thompson.java
package PLD.classes;
import java.io.*;
import java.util.*;

public class Thompson {
    private Stack<FA> nfas = new Stack<>();
    private String regex;
    private Lexer lexer;
    private Parser parser;
    public static final char EPSILON = '\u03F5';

    public Thompson(String regex) {
        this.regex = regex;
        this.lexer = new Lexer(regex);
        this.parser = new Parser(lexer);
    }

    public void baseStep(char character) {
        ArrayList<State> states = new ArrayList<>();
        ArrayList<Transition> transitions = new ArrayList<>();
        Transition transition = new Transition();

        transition.addTransition("s1", character);
        transitions.add(transition);

        State initial_state = new State("s0", transitions, false, true);
        State final_state = new State("s1", new ArrayList<>(), true, false);

        states.add(initial_state);
        states.add(final_state);

        nfas.push(new FA(new ArrayList<>(), states));
    }

    private void OR(FA A, FA B){
        ArrayList<State> states = new ArrayList<>();
        ArrayList<String> alphabet = new ArrayList<>();
        int numberOfStates = A.getStatesNumber() + B.getStatesNumber();

        ArrayList<Transition> initTransitions = new ArrayList<>();
        initTransitions.add(new Transition("s1", EPSILON));
        initTransitions.add(new Transition("s" + (B.getStatesNumber() + 1), EPSILON));
        State initial_state = new State("s0", initTransitions, false, true);
        states.add(initial_state);

        for(State state : B.getStates()) {
            setORStates(states, numberOfStates, state, 0);
        }

        for(State state : A.getStates()) {
            setORStates(states, numberOfStates, state, B.getStates().size());
        }

        State final_state = new State("s" + (numberOfStates + 1), new ArrayList<>(), true, false);
        states.add((final_state));
        nfas.push(new FA(new ArrayList<>(), states));
    }

    private void setORStates(ArrayList<State> states, int numberOfStates, State state, int extraValue) {
        int stateNumber = Integer.parseInt(String.valueOf(state.getId().charAt(1)));
        state.setId("s" + (stateNumber + 1 + extraValue));
        state.setInitial(false);
        state.setFinal(false);

        if(state.getTransitions().size() == 0){
            ArrayList<Transition> tempTransitions = new ArrayList<>();
            Transition tempTransition1 = new Transition();
            tempTransition1.addTransition("s" + (numberOfStates + 1), EPSILON);
            tempTransitions.add(tempTransition1);
            state.setTransitions(tempTransitions);
        }else{
            for(Transition transition: state.getTransitions()){
                int transitionNumber = Integer.parseInt(String.valueOf(transition.getKey().charAt(1)));
                transition.setKey("s" + (transitionNumber + 1 + extraValue));
            }
        }
        states.add(state);
    }

    /**
     * Executes the KLEAN operation
     * @param A where A is an FA
     */
    private void KLEAN(FA A){
        //TODO
        ArrayList<State> states = new ArrayList<>();
        ArrayList<String> alphabet = new ArrayList<>();
        int numberOfStates = A.getStatesNumber();

        ArrayList<Transition> transitions = new ArrayList<>();
        transitions.add(new Transition("s1", EPSILON));
        transitions.add(new Transition("s" + (numberOfStates + 1), EPSILON));
        State initial_state = new State("s0", transitions, false, true);

        states.add(initial_state);


        for(State state : A.getStates()){
            int stateNumber = Integer.parseInt(String.valueOf(state.getId().charAt(1)));
            state.setId("s" + (stateNumber + 1));
            state.setInitial(false);
            state.setFinal(false);

            if(state.getTransitions().size() == 0){
                ArrayList<Transition> tempTransitions = new ArrayList<>();
                Transition tempTransition1 = new Transition();
                Transition tempTransition2 = new Transition();
                tempTransition1.addTransition("s" + (numberOfStates + 1), EPSILON);
                tempTransition2.addTransition("s1", EPSILON);
                tempTransitions.add(tempTransition1);
                tempTransitions.add(tempTransition2);
                state.setTransitions(tempTransitions);
            }else{
                for(Transition transition: state.getTransitions()){
                    int transitionNumber = Integer.parseInt(String.valueOf(transition.getKey().charAt(1)));
                    transition.setKey("s" + (transitionNumber + 1));
                }
            }

            states.add(state);
        }
        State final_state = new State("s" + (numberOfStates + 1), new ArrayList<>(), true, false);
        states.add((final_state));
        nfas.push(new FA(new ArrayList<>(), states));
    }

    /**
     * Executes the KLEAN operation
     * @param A where A is a FA
     * @param B where B is a FA
     */
    private void CONCAT(FA A, FA B){
        //TODO
        ArrayList<State> states = new ArrayList<>();
        ArrayList<String> alphabet = new ArrayList<>();
        int numberOfStates = B.getStatesNumber() - 1;

        State a = B.getStates().get(B.getStates().size() - 1);
        State b = A.getStates().get(0);
        a.setFinal(false);
        b.setInitial(false);

        states.addAll(B.getStates());
        states.remove(states.size() - 1);

        for(State state: A.getStates()){
            for(Transition transition: state.getTransitions()){
                int newPosition = Integer.parseInt(String.valueOf(transition.getKey().charAt(1))) + numberOfStates;
                transition.setKey("s" + newPosition);
            }

            if(state.getId().equals("s0")){
                state.setId("s" + numberOfStates);
            }else {
                state.setId("s" + (Integer.parseInt(String.valueOf(state.getId().charAt(1))) + numberOfStates));
            }


            states.add(state);
        }
        nfas.push(new FA(new ArrayList<>(), states));
    }

    //TODO
    private void PLUS(FA A){
        FA B = new FA(A);
        KLEAN(A);
        CONCAT(this.nfas.pop(), B);
    }

    public FA executeAlgorithm() throws IOException {
        this.regex = parser.parse();
        for (int i = 0; i < this.regex.length(); i++) {
            char character = this.regex.charAt(i);
            switch (character) {
                case '*' -> KLEAN(this.nfas.pop());
                case '+' -> PLUS(this.nfas.pop());
                case '?' -> {
                    baseStep(EPSILON);
                    OR(this.nfas.pop(), this.nfas.pop());
                }
                case '&' -> CONCAT(this.nfas.pop(), this.nfas.pop());
                case '|' -> OR(this.nfas.pop(), this.nfas.pop());
                default -> baseStep(character);
            }
        }

        FA finalSM = this.nfas.pop();
        finalSM.graph(finalSM, false);
        return finalSM;
    }
}

File: Project_2/classes/Thompson.py
from classes.FA import FA
from classes.Lexer import Lexer
from classes.Parser import Parser
from classes.State import State
from classes.Transition import Transition


class Thompson:
    def __init__(self, regex):
        self.regex = regex
        self.lexer = Lexer(regex)
        self.parser = Parser(self.lexer)
        self.fas = []
        self.EPSILON = '\u03F5'

    def base_step(self, character):
        states = []
        transitions = []
        transition = Transition()

        transition.add_transition("s1", character)
        transitions.append(transition)

        initial_state = State("s0", transitions, False, True)
        final_state = State("s1", [], True, False)

        states.append(initial_state)
        states.append(final_state)

        self.fas.append(FA([], states))

    def __OR(self, A, B):
        states = []
        alphabet = []
        number_of_states = len(A.states) + len(B.states)

        init_transitions = [Transition("s1", self.EPSILON), Transition("s" + (len(B.states)), self.EPSILON)]

        initial_state = State("s0", init_transitions, False, True)
        states.append(initial_state)

        for state in B.states:
            self.__set_OR_states(states, number_of_states, state)

        for state in A.states:
            self.__set_OR_states(states, number_of_states, state, len(B.states))

        final_state = State("s" + (number_of_states + 1), [], True, False)
        states.append(final_state)
        self.fas.append(FA([], states))

    def __set_OR_states(self, states, number_of_states, state, extra_value=0):
        state_number = int(state.id[1])
        state.id = "s" + str((state_number + 1 + extra_value))
        state.is_initial = False
        state.is_final = False

        if len(state.transitions) == 0:
            temp_transitions = []
            temp_transition_1 = Transition()
            temp_transition_1.add_transition("s" + (number_of_states + 1), self.EPSILON)
            temp_transitions.append(temp_transition_1)
            state.transitions = temp_transitions
        else:
            for transition in state.transitions:
                transition_number = int(transition.key[1])
                transition.key = "s" + str((transition_number + 1 + extra_value))

        states.append(state)

    def __KLEAN(self, A):
        states = []
        number_of_states = len(A.states)

        transitions = [Transition("s1", self.EPSILON), Transition("s" + str((number_of_states + 1)), self.EPSILON)]
        initial_state = State("s0", transitions, False, True)

        states.append(initial_state)

        for state in A.states:
            state_number = int(state.id[1])
            state.id = "s" + (state_number + 1)
            state.is_initial = False
            state.is_final = False

            if len(state.transitions) == 0:
                temp_transitions = []
                temp_transition_1 = Transition()
                temp_transition_2 = Transition()
                temp_transition_1.add_transition("s" + str((number_of_states + 1)), self.EPSILON)
                temp_transition_2.add_transition("s1", self.EPSILON)
                temp_transitions.append(temp_transition_1)
                temp_transitions.append(temp_transition_2)
                state.transitions = temp_transitions

            else:
                for transition in state.transitions:
                    transition_number = int(transition.key[1])
                    transition.key = "s" + str((transition_number + 1))

            states.append(state)

        final_state = State("s" + str((number_of_states + 1)), [], True, False)
        states.append(final_state)
        self.fas.append(FA([], states))

    def __CONCAT(self, A, B):
        states = []
        number_of_states = int(len(B.states) - 1)

        a = B.states[len(B.states) - 1]
        b = A.states[0]

        states.append(B.states)
        states.pop(len(states) - 1)

        for state in A.states:
            for transition in state.transitions:
                new_position = int(transition.key[1]) + int(number_of_states)
                transition.key = "s" + str(new_position)

            if state.id == "s0":
                state.id = "s" + str(number_of_states)
            else:
                state.id = "s" + str(int((int(state.id[1]) + int(number_of_states))))

            states.append(state)

        self.fas.append(FA([], states))

    def __PLUS(self, A):
        B = FA(A)
        self.__KLEAN(A)
        self.__CONCAT(self.fas.pop(), B)

    def execute_algorithm(self):
        self.regex = self.parser.parse()

        for i in range(len(self.regex)):
            character = self.regex[i]

            if character == '*':
                self.__KLEAN(self.fas.pop())
            elif character == '+':
                self.__PLUS(self.fas.pop())
            elif character == '?':
                self.base_step(self.EPSILON)
                self.__OR(self.fas.pop(), self.fas.pop())
            elif character == '&':
                self.__CONCAT(self.fas.pop(), self.fas.pop())
            elif character == '|':
                self.__OR(self.fas.pop(), self.fas.pop())
            else:
                self.base_step(character)
        final_sm = self.fas.pop()
        final_sm.graph(final_sm, False)

        return final_sm